### PR TITLE
Fix tree kinds

### PR DESCRIPTION
**Pull request by jyluo**
draft fix

DO NOT MERGE

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/util/element/ElementAnnotationUtil.java [hunk 1]
```diff
@@ -435,6 +454,19 @@ private static AnnotatedTypeMirror getLocationTypeADT(
             enclosing = enclosing.getEnclosingType();
         }
 
+        int numOfEnclosingTypes = outerToInner.size() - 1;
+
+        // If the AnnotatedDeclaredType is a component of an array type, then apply anno to all
+        // inner types of the given declared type if the declared type has enclosing classes
```
Comment (jyluo) [inline framework/src/main/java/org/checkerframework/framework/util/element/ElementAnnotationUtil.java]: "to all possible inner types."

Add link to https://bugs.openjdk.java.net/browse/JDK-8208470 and state that this fix can be removed once JDK bug is fixed

File: framework/src/main/java/org/checkerframework/framework/util/element/TypeVarUseApplier.java [hunk 2]
```diff
@@ -170,7 +170,8 @@ public void extractAndApply() {
 
     private boolean isBaseComponent(
             final AnnotatedArrayType arrayType, final Attribute.TypeCompound anno) {
-        return ElementAnnotationUtil.getTypeAtLocation(arrayType, anno.getPosition().location)
+        return ElementAnnotationUtil.getTypeAtLocation(
+                        arrayType, anno.getPosition().location, anno, false)
```
Comment (jyluo) [inline framework/src/main/java/org/checkerframework/framework/util/element/TypeVarUseApplier.java]: pass null for anno, and check that addAnnotation handles null input (should do nothing)

**Comment by jyluo** [comment]
Given test code
```
List<@TA(1) ? extends @TA(2) Inner []> lstBothBound;
List<@TA(1) ? extends ArrayOfInner.@TA(2) Inner []> lstBothExplicitBound;
```
javac 8 and 9 gives us locations
@TA(1): FIELD, location=[TYPE_ARGUMENT(0)]
@TA(2): FIELD, location=[TYPE_ARGUMENT(0), WILDCARD, ARRAY, INNER_TYPE]
for both of these field, so it seems like it correctly generates the location byte code for array types used as bounds on wildcards.

I've changed `getLocationTypeAWT()` to pass null and false in its recursive calls of getTypeAtLocation since we don't need special handling of this case.

**Comment by wmdietl** [comment]
Thanks for the update!
How do the error numbers look before/after this change?

**Comment by jyluo** [comment]
The numbers are identical. No new bugs introduced, and same error fixed as before.

File: framework/src/main/java/org/checkerframework/framework/util/element/ElementAnnotationUtil.java [hunk 3]
```diff
@@ -451,7 +482,8 @@ private static AnnotatedTypeMirror getLocationTypeADT(
                     if (currentLocation.arg < innerType.getTypeArguments().size()) {
                         AnnotatedTypeMirror typeArg =
                                 innerType.getTypeArguments().get(currentLocation.arg);
-                        return getTypeAtLocation(typeArg, tailOfLocations);
+                        return getTypeAtLocation(
+                                typeArg, tailOfLocations, anno, isComponentTypeOfArray);
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/element/ElementAnnotationUtil.java]: Is this necessary in a type argument position? I thought only top-level annos are an issue?

File: framework/src/main/java/org/checkerframework/framework/util/element/ElementAnnotationUtil.java [hunk 4]
```diff
@@ -499,11 +531,11 @@ private static AnnotatedTypeMirror getLocationTypeAWT(
         if (!location.isEmpty()
                 && location.get(0).tag.equals(TypeAnnotationPosition.TypePathEntryKind.WILDCARD)) {
             if (AnnotatedTypes.hasExplicitExtendsBound(type)) {
-                return getTypeAtLocation(type.getExtendsBound(), tail(location));
+                return getTypeAtLocation(type.getExtendsBound(), tail(location), null, false);
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/element/ElementAnnotationUtil.java]: How about adding an overload of `getTypeAtLocation` that passes `null, false`.
Then most invocations won't need to be changed and the few that need change can be understood and documented.

**Comment by wmdietl** [comment]
Thanks, I've simplified the code a bit and merged this.

**Pull request merged**